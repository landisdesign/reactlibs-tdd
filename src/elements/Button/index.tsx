import React from 'react';
import styled from 'styled-components';
import { colors, fontWeight, fontSize, fontFamily, media } from '../../common/styling';

/**
 * Properties for Button, in addition to those available to HTMLButtonElement
 *
 *  @property isDefault Indicate this the default button in a list of buttons
 *  @property render If provided, overrides `children` as the content for the button
 *  @property className if provided, overrides default styling
 */
export interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
    default?: boolean;
    render?: () => React.ReactNode;
}

const Button: React.FC<ButtonProps> = (props) => {

    if (props.className) {
        const {
            default: isDefault,
            render,
            children,
            ...htmlProps
        } = props;
        return <button {...htmlProps}>{render ? render() : children}</button>;
    }

    const {
        render,
        children
    } = props;
    return <StyledButton {...props}>{render ? render() : children}</StyledButton>;
}

export default Button;

const gradientValues = {
    plain: {
        range: [5, 10, 90, 100],
        accented: [colors.canvas, colors.accentPale, colors.accentMedium, colors.accentDark],
        standard: [colors.canvas, () => '#EEE', () => '#BBB', () => '#888']
    },
    active: {
        range: [0, 10, 90, 100],
        accented: [colors.accentDark, colors.accentPale, colors.accentMedium, colors.canvas],
        standard: [() => '#888', () => '#EEE', () => '#BBB', colors.canvas]
    }
};

const backgroundGradient = (isActive: boolean) => (props: ButtonProps) => {
    const {
        range,
        accented,
        standard
    } = gradientValues[isActive ? 'active' : 'plain'];

    let colorSet = props.default ? accented : standard;
    let steps = [];
    for (const i in range) {
        steps.push(`${colorSet[i](props)} ${range[i]}%`);
    }
    return `linear-gradient(to bottom, ${steps.join(', ')})`;
}

const StyledButton = styled.button<ButtonProps>`

    background: ${backgroundGradient(false)};
    border: 1px solid ${colors.accentMedium};
    border-radius: .25rem;
    white-space: nowrap;
    cursor: pointer;

    &:hover {
        background: ${props => props.default ? colors.accentLight(props) : '#DDD'};
    }

    &:active {
        background: ${backgroundGradient(true)};
    }

    &[disabled] {
        border: 1px solid #CCC;
        background: ${props => props.default ? '#DDD' : '#EEE'};
        color: #666;
        cursor: arrow;
    }

    ${media.desktop`
        font: ${fontWeight.bold} ${fontSize.desktop.small} ${fontFamily.sans};
    `}

    ${media.notDesktop`
        font: ${fontWeight.bold} ${fontSize.tablet.small} ${fontFamily.sans};
    `}
`;

