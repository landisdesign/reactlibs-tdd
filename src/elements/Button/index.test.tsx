import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Button from '.';
import { ThemeProvider } from 'styled-components';
import 'jest-styled-components';

const getButton = (button: React.ReactNode) =>
    mount(<ThemeProvider theme={{ mode: 'light' }}>{button}</ThemeProvider>).find('button');

test('Standard <button> attributes passed through', () => {
    const testClick = jest.fn();
    let button = getButton(<Button aria-pressed='false' type='submit' onClick={testClick}>click</Button>);

    expect(button.text()).toEqual('click');
    expect(button.prop('aria-pressed')).toEqual('false');
    expect(button.prop('type')).toEqual('submit');
    button.simulate('click');
    expect(testClick.mock.calls).toHaveLength(1);

    button = getButton(<Button disabled>disabled</Button>);
    expect(button.prop('disabled')).toBeDefined();
});

test('Button outputs correct styling', () => {
    let button = getButton(<Button>standard</Button>);
    expect(toJson(button)).toMatchSnapshot();

    button = getButton(<Button disabled>plain disabled</Button>);
    expect(toJson(button)).toMatchSnapshot();

    button = getButton(<Button default>default</Button>);
    expect(toJson(button)).toMatchSnapshot();

    button = getButton(<Button disabled default>default disabled</Button>);
    expect(toJson(button)).toMatchSnapshot();
});

test('class names override initial classes', () => {
    let button = getButton(<Button className='different-class'>different</Button>);
    expect(button.prop('className')).toEqual('different-class');

    button = getButton(<Button default className='another-class'>another</Button>);
    expect(button.prop('className')).toEqual('another-class');
});

test('children rendered inside <button>', () => {
    const content = <b>Content</b>;
    const button = getButton(<Button>{ content }</Button>);
    expect(button.contains(content)).toBe(true);
});

test('render function overrides children', () => {
    const content = <b>Content</b>;
    const button = getButton(<Button render={ () => content }/>);
    expect(button.contains(content)).toBe(true);
});
