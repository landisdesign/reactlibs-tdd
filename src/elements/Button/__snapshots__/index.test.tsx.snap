// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button outputs correct styling 1`] = `
.c0 {
  background: linear-gradient(to bottom,#FFF 5%,#EEE 10%,#BBB 90%,#888 100%);
  border: 1px solid #689;
  border-radius: .25rem;
  white-space: nowrap;
  cursor: pointer;
}

.c0:hover {
  background: #DDD;
}

.c0:active {
  background: linear-gradient(to bottom,#888 0%,#EEE 10%,#BBB 90%,#FFF 100%);
}

.c0[disabled] {
  border: 1px solid #CCC;
  background: #EEE;
  color: #666;
  cursor: arrow;
}

@media (min-width:48rem) {
  .c0 {
    font: 900 1.25rem "mr-eaves-modern",sans-serif;
  }
}

@media (max-width:48rem) {
  .c0 {
    font: 900 .85rem "mr-eaves-modern",sans-serif;
  }
}

<button
  className="c0"
>
  standard
</button>
`;

exports[`Button outputs correct styling 2`] = `
.c0 {
  background: linear-gradient(to bottom,#FFF 5%,#EEE 10%,#BBB 90%,#888 100%);
  border: 1px solid #689;
  border-radius: .25rem;
  white-space: nowrap;
  cursor: pointer;
}

.c0:hover {
  background: #DDD;
}

.c0:active {
  background: linear-gradient(to bottom,#888 0%,#EEE 10%,#BBB 90%,#FFF 100%);
}

.c0[disabled] {
  border: 1px solid #CCC;
  background: #EEE;
  color: #666;
  cursor: arrow;
}

@media (min-width:48rem) {
  .c0 {
    font: 900 1.25rem "mr-eaves-modern",sans-serif;
  }
}

@media (max-width:48rem) {
  .c0 {
    font: 900 .85rem "mr-eaves-modern",sans-serif;
  }
}

<button
  className="c0"
  disabled={true}
>
  plain disabled
</button>
`;

exports[`Button outputs correct styling 3`] = `
.c0 {
  background: linear-gradient(to bottom,#FFF 5%,#DEF 10%,#689 90%,#356 100%);
  border: 1px solid #689;
  border-radius: .25rem;
  white-space: nowrap;
  cursor: pointer;
}

.c0:hover {
  background: #9AC;
}

.c0:active {
  background: linear-gradient(to bottom,#356 0%,#DEF 10%,#689 90%,#FFF 100%);
}

.c0[disabled] {
  border: 1px solid #CCC;
  background: #DDD;
  color: #666;
  cursor: arrow;
}

@media (min-width:48rem) {

}

@media (max-width:48rem) {

}

@media (min-width:48rem) {
  .c0 {
    font: 900 1.25rem "mr-eaves-modern",sans-serif;
  }
}

@media (max-width:48rem) {
  .c0 {
    font: 900 .85rem "mr-eaves-modern",sans-serif;
  }
}

<button
  className="c0"
  default={true}
>
  default
</button>
`;

exports[`Button outputs correct styling 4`] = `
.c0 {
  background: linear-gradient(to bottom,#FFF 5%,#DEF 10%,#689 90%,#356 100%);
  border: 1px solid #689;
  border-radius: .25rem;
  white-space: nowrap;
  cursor: pointer;
}

.c0:hover {
  background: #9AC;
}

.c0:active {
  background: linear-gradient(to bottom,#356 0%,#DEF 10%,#689 90%,#FFF 100%);
}

.c0[disabled] {
  border: 1px solid #CCC;
  background: #DDD;
  color: #666;
  cursor: arrow;
}

@media (min-width:48rem) {

}

@media (max-width:48rem) {

}

@media (min-width:48rem) {
  .c0 {
    font: 900 1.25rem "mr-eaves-modern",sans-serif;
  }
}

@media (max-width:48rem) {
  .c0 {
    font: 900 .85rem "mr-eaves-modern",sans-serif;
  }
}

<button
  className="c0"
  default={true}
  disabled={true}
>
  default disabled
</button>
`;
